### DISCUSSION TITLE
Remote Function Calls from Dirigible

### DISCUSSION THREAD

**Discussion by vwetter** [eclipse/dirigible#1955] [Q&A]
Hi,
is it possible to perform a Remote Function Call (RFC) against an ABAP stack (within the same network) ? If yes, what is required (e.g. SAP JCo, RFC SDK, Cloud SDK) ? I wasn't able to find any information for this.
The Dirigible App should only act as RFC client, not RFC server.

Thanks

**Comment by ThuF** [comment]
Hi @vwetter,

Yes, it's possible to perform RFC, but you would need SAP JCo.
To do so, you'll need to download the SAP JCo and build your own distribution with it (e.g. Docker image):

**Dockerfile:**
```
FROM dirigiblelabs/dirigible-base-platform:6.3.12

COPY <SAP-JCo-JAR-Files-Location>/<sap-jco-jar1>.jar $CATALINA_HOME/webapps/ROOT/WEB-INF/lib

EXPOSE 8080 8081

CMD ["catalina.sh", "jpda", "run"]
```

As I'm not that familiar with the SAP JCo, I guess this Maven Dependency could be also used to package it: https://mvnrepository.com/artifact/com.sap.conn.jco/com.sap.conn.jco

After you've packaged the SAP JCo dependency, then you could call the JCo Java classes by FQN:

```javascript
let destination = com.sap.conn.jco.JCoDestinationManager.getDestination("myDestination");
let ft = destination.getRepository().getFunctionTemplate("bapiName");
let myFunction = ft.getFunction();
...
```

Here you could find how to build a custom stack: https://www.dirigible.io/samples/tutorials/customizations/custom-stack/

Also you could reach out to us via Slack: https://slack.dirigible.io

Regards,
Yordan

**Reply by vwetter** [reply]
Hi Yordan,
fyi, I got it working last night as a container on a Windows 10 docker host. Simple call to an ABAP stack (RFC_SYSTEM_INFO) works now. Thanks again for your help! 

Volker

**Reply by ThuF** [reply]
Hi Volker,

That's great!
Which approach you've used?
Could you share a bit more details (about the setup), so we could make a documentation out of it :)

Btw, as far as I know, it should be possible to expose RFC functions via OData or SOAP WebServices:
- https://blogs.sap.com/2012/10/26/step-by-step-guide-to-build-an-odata-service-based-on-rfcs-part-1/

If the OData/WebServices exposure works for you, then you could leverage the HTTP Client or the SOAP APIs:
- https://www.dirigible.io/api/http/client/
- https://www.dirigible.io/api/net/soap/

Regards,
Yordan

**Comment by ThuF** [comment]
@vwetter 

You can also check out these tutorials:
- https://www.dirigible.io/samples/tutorials/customizations/custom-stack/facade/
- https://www.dirigible.io/samples/tutorials/customizations/custom-stack/dependency/

**Comment by ThuF** [comment]
Here you could find how to do it on the SAP BTP Cloud Foundry environment:

- https://blogs.sap.com/2020/07/30/sap-cloud-foundry-to-on-premise-rfc-connection-using-jco/

**Reply by vwetter** [reply]
Many thanks, Yordan. Very helpful information. I‘ll give it a try.
